### PR TITLE
Add gtk composite template support v2

### PR DESCRIPTION
**Pull request by zii-dmg**
Second version of template support.
Most code is from https://github.com/GtkSharp/GtkSharp/pull/90.
Auto signal connections are working, see samples.
Also added check for unknown fields (when GTK can't find widget in .glade).

But no .NET project templates, sorry.

### CODE REVIEW

File: Source/Libs/GtkSharp/Widget.cs [hunk 1]
```diff
@@ -232,13 +255,109 @@ static void ClassInit (GLib.GType gtype, Type t)
...
 			CssNameAttribute attr = t.GetCustomAttribute<CssNameAttribute>(true);
 			if (attr != null)
 				SetCssName (gtype, attr.Name);
 		}
 
+		protected Widget () : base (IntPtr.Zero)
+		{
+			CreateNativeObject (Array.Empty<string> (), Array.Empty<GLib.Value> ());
+		}
+
```
Comment (lytico) [inline Source/Libs/GtkSharp/Widget.cs:328]: why this change is needed?

the autogenerated code looks like this:

```
		protected Widget() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```
it's the same, except modernizing empty array

File: Source/Libs/GtkSharp/Bin.cs [hunk 2]
```diff
@@ -24,6 +24,9 @@ namespace Gtk {
 	using System.Runtime.InteropServices;
 
 	public partial class Bin {
+
+		protected Bin() : base() { }
+
```
Comment (lytico) [inline Source/Libs/GtkSharp/Bin.cs:29]: why this change is needed?

the autogenerated code looked like this:

```
		protected Bin() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```

File: Source/Libs/GtkSharp/Box.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Gtk {
+
+	using System;
+	using System.Runtime.InteropServices;
+
+	public partial class Box {
+
+		protected Box() : base() { }
+	}
```
Comment (lytico) [inline Source/Libs/GtkSharp/Box.cs:29]: why this change is needed?

the autogenerated code looked like this:

```
		protected Box() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```

File: Source/Libs/GtkSharp/Container.cs [hunk 4]
```diff
@@ -51,6 +51,8 @@ struct GParamSpec
             public uint param_id;
         };
 
+        protected Container() : base() { }
```
Comment (lytico) [inline Source/Libs/GtkSharp/Container.cs:54]: hy this change is needed?

the autogenerated code looked like this:

```
		protected Container() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```

File: Source/Libs/GtkSharp/Bin.cs (hunk 2 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Bin.cs:29]: Just to minimize memory allocations

File: Source/Libs/GtkSharp/Box.cs (hunk 3 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: Don't know, taken from #90

File: Source/Libs/GtkSharp/Container.cs (hunk 4 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Container.cs:54]: Don't know, taken from #90

File: Source/Libs/GtkSharp/Widget.cs (hunk 1 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs:328]: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825

Comment (lytico) [reply Source/Libs/GtkSharp/Widget.cs:328]: that's nice and good, but should't we do this generally in the code generator?

Comment (lytico) [reply Source/Libs/GtkSharp/Widget.cs:328]: i copied from develop branch:

```
#region Autogenerated code
	[GLib.TypeInitializer (typeof (Gtk.Widget), "ClassInit")]
	public partial class Widget : GLib.InitiallyUnowned, Atk.IImplementor {

		public Widget (IntPtr raw) : base(raw) {}

		protected Widget() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```
what i'm missing?

Comment (lytico) [reply Source/Libs/GtkSharp/Widget.cs:328]: btw, i'm currently reviewing generator code to fix that generally.

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs:328]: You mean that generated code is same? So we don't need `<attr path="/api/namespace/object[@cname='GtkWidget']" name="disable_void_ctor">true</attr>` ?

Comment (lytico) [reply Source/Libs/GtkSharp/Widget.cs:328]: yes, if we change to
 `Array.Empty<string> (), Array.Empty<GLib.Value> () `
in code generator.

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs:328]: I agree

Comment (lytico) [reply Source/Libs/GtkSharp/Widget.cs:328]: so i change code generator with a separate PR, and we can revert changes in this pr concerning `disable_void_ctor` in api

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs:328]: Yes

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs:328]: Good

File: Source/Libs/GtkSharp/Box.cs (hunk 3 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: For `Box` protected constructor is not generated

File: Source/Libs/GtkSharp/Container.cs (hunk 4 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Container.cs]: Reverted

File: Source/Libs/GtkSharp/Bin.cs (hunk 2 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Bin.cs]: Reverted

File: Source/Libs/GtkSharp/Widget.cs (hunk 1 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Widget.cs]: Reverted

**Comment by zii-dmg** [comment]
Don't know about nuget CI errors, but build is ok

File: Source/Libs/GtkSharp/Box.cs (hunk 3 above)
Comment (lytico) [reply Source/Libs/GtkSharp/Box.cs:29]: box again ...
before, there was no chance to subclass a Box without base constructor 

`public Box (Gtk.Orientation orientation, int spacing) : base (IntPtr.Zero)`

now, i can subclass like this:

```
        public class MyBox : Box {
            
        }
```

i don't know if this is intended ....

**Comment by lytico** [comment]
> Don't know about nuget CI errors, but build is ok

same here ... https://github.com/GtkSharp/GtkSharp/pull/395

File: Source/Libs/GtkSharp/Box.cs (hunk 3 above)
Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: Intended. Now you can create template for `Box`. So `Box` is root class in glade.

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: Glade is not very friendly for Box, but you can create Box. Select Box in Containers popup, then right click in object tree and `Add widget as toplevel`.

Comment (lytico) [reply Source/Libs/GtkSharp/Box.cs:29]: despite this...:

https://docs.gtk.org/gtk3/ctor.Box.new.html

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: https://pygobject.readthedocs.io/en/latest/guide/gtk_template.html

In python they inherit from Box also for template example

Comment (lytico) [reply Source/Libs/GtkSharp/Box.cs:29]: i tried MyBox in samples, and it seems to work without error with simple usage.
but i don't know if something wents wrong with more complex usages.

Should we investigate more???

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: I think current behavior is good. Can be fixed later if something is wrong.

Comment (lytico) [reply Source/Libs/GtkSharp/Box.cs:29]: with this implementation 
`gtk_box_new`
is never called.

Comment (zii-dmg) [reply Source/Libs/GtkSharp/Box.cs:29]: `gtk_box_new` just sets `orientation` and `spacing` properties but they already have initialized values. C# class `Object.CreateNativeObject` -> `g_object_newv`.

https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gtk/gtkbox.c

```c
GtkWidget*
gtk_box_new (GtkOrientation orientation,
             gint           spacing)
{
  return g_object_new (GTK_TYPE_BOX,
                       "orientation", orientation,
                       "spacing",     spacing,
                       NULL);
}
```

```c
static void
gtk_box_init (GtkBox *box)
{
  GtkBoxPrivate *private;

  box->priv = gtk_box_get_instance_private (box);
  private = box->priv;

  gtk_widget_set_has_window (GTK_WIDGET (box), FALSE);

  private->orientation = GTK_ORIENTATION_HORIZONTAL;
  private->children = NULL;

  private->default_expand = FALSE;
  private->homogeneous = FALSE;
  private->spacing = 0;
  private->spacing_set = FALSE;
  ...
```

Comment (lytico) [reply Source/Libs/GtkSharp/Box.cs:29]: i see ... so we should have no problems, as the calls are  identical in the end.

File: Source/Libs/GtkSharp/Container.cs [hunk 5]
```diff
@@ -80,7 +80,7 @@ public IEnumerator GetEnumerator()
 
         class ChildAccumulator
         {
-            public ArrayList Children = new ArrayList();
+            public System.Collections.Generic.List<Widget> Children = new System.Collections.Generic.List<Widget>();
```
Comment (lytico) [inline Source/Libs/GtkSharp/Container.cs:83]: code modernization; should be ok

File: Source/Libs/GtkSharp/Container.cs [hunk 6]
```diff
@@ -97,12 +97,12 @@ public IEnumerable AllChildren
                 return acc.Children;
             }
         }
-		[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-		delegate bool d_gtk_container_get_focus_chain(IntPtr raw, out IntPtr list_ptr);
-		static d_gtk_container_get_focus_chain gtk_container_get_focus_chain = FuncLoader.LoadFunction<d_gtk_container_get_focus_chain>(FuncLoader.GetProcAddress(GLibrary.Load(Library.Gtk), "gtk_container_get_focus_chain"));
-		[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-		delegate void d_gtk_container_set_focus_chain(IntPtr raw, IntPtr list_ptr);
-		static d_gtk_container_set_focus_chain gtk_container_set_focus_chain = FuncLoader.LoadFunction<d_gtk_container_set_focus_chain>(FuncLoader.GetProcAddress(GLibrary.Load(Library.Gtk), "gtk_container_set_focus_chain"));
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
```
Comment (lytico) [inline Source/Libs/GtkSharp/Container.cs:100]: code reformat

File: Source/Libs/GtkSharp/HandlerNotFoundException.cs
Comment (lytico) [inline Source/Libs/GtkSharp/HandlerNotFoundException.cs:1]: code cleanup; moved from Builder.cs

File: Source/Libs/GtkSharp/SignalConnector.cs
Comment (lytico) [inline Source/Libs/GtkSharp/SignalConnector.cs:1]: code cleanup; moved from Builder

**Comment by lytico** [comment]
examples tested; working

**Comment by lytico** [comment]
check: build is ok, just nuget error

**Comment by lytico** [comment]
@zii-dmg thanks!
